wpengine/atlas-content-modeler
# [BH-1123] Prevent use of reserved slugs during field creation

## Description

Fixes #312 by preventing the use of field slugs with reserved keywords. Works by:

• Using a list of reserved field names used by GraphQL for the default WP Post type (client-side and server-side) — see `includes/settings/reserved-field-slugs.php`.
• Using GraphQL introspection to identify additional fields registered at runtime if WPGraphQL is available (client-side) — see the new `useReservedSlugs` custom hook in `includes/settings/js/src/hooks/useReservedSlugs.js`.

https://wpengine.atlassian.net/browse/BH-1123

## Testing

• Includes unit test to check that fields created via the REST endpoint cannot use default reserved field names.
• Includes e2e test for reserved field names registered at runtime.

### To test reserved default words manually

1. Click '+' to add a new field.
2. Give the field a name of 'id'.
3. Tab to the API Identifier field.

You should see a message that the 'id' slug is reserved or in use.

### To test reserved fields registered by another plugin at runtime

The PR will prevent WPGraphQL field names such as 'seo' from being used as field names if another plugin registered them with WPGraphQL:

1. Install and activate _Yoast SEO_ and the _WPGraphQL Yoast SEO Addon_: `wp plugin install --activate wordpress-seo add-wpgraphql-seo`
2. Click '+' to add a new field.
3. Give the field a name of 'seo' (this is a field name registered with WPGraphQL by the WPGraphQL Yoast SEO Addon at run-time.)
4. Tab to the API Identifier field.

You should see a message that the 'seo' slug is reserved or in use.

## Screenshots

https://user-images.githubusercontent.com/647669/137753474-f4cdcbf9-784d-4d75-9cf5-5194128957e0.mp4

## Documentation Changes

n/a

## Dependant PRs

n/a
